### PR TITLE
Fix wrong results from npartitions

### LINKED ISSUE #24267: npartitions gives incorrect results
```npartitions(n)``` gives incorrect results for n=1001, 1055, 1077, 1110, 1111, ... . The results are off by one, sometimes up, sometimes down. For example, ```npartitions(1001)``` gives 25032297938763929621013218349795 but the correct value ought to be 25032297938763929621013218349796. I checked against Sage, Pari, [OEIS](https://oeis.org/A000041), and [WolframAlpha](https://www.wolframalpha.com/input?i=partitions+of+1001).

Testing in [SageMathCell](https://sagecell.sagemath.org/?z=eJxtzLEKgzAQxvFd8B2ukwm0JTgK9ywhpTm5IZdwpoOI726x1slv-E8_Pk4la4VpTmVuG8oKAiygQcZoeuecHdoGvit-N7j3KSVo5cpZJiP2FGGMKJ_0iuoz-QvD9D-64c8PUJSlGuoEF1nv8GYiXA71OFC_dnYDHf85yQ==&lang=sage&interacts=eJyLjgUAARUAuQ==) with
```
import sympy
for n in range(2000):
    p_sympy=sympy.npartitions(n)
    p_sage=number_of_partitions(n)
    if p_sympy!=p_sage: print(f'n={n}, diff={p_sympy-p_sage:2}')
```
gives the output
```
n=1001, diff=-1
n=1055, diff= 1
n=1077, diff= 1
n=1110, diff= 1
n=1111, diff=-1
n=1259, diff= 1
n=1339, diff=-1
n=1373, diff=-1
n=1623, diff= 1
n=1636, diff= 1
n=1637, diff= 1
n=1662, diff= 1
n=1701, diff= 1
n=1703, diff=-1
n=1715, diff= 1
n=1729, diff=-1
n=1889, diff= 1
n=1903, diff= 1
```

### PR DESCRIPTION
**Pull request by maciejskorski**
#### References to other Issues or PRs

Fixes #24267

#### Brief description of what is fixed or changed

Takes more terms in the series approximation and increases the arithmetic precision.

#### Other comments

[x] work out the right precision.

#### Release Notes

* ntheory
  * fixes the numerical evaluation of `npartitions`.

### CODE REVIEW

File: sympy/ntheory/partitions_.py [hunk 1]
```diff
@@ -52,7 +52,6 @@ def _a(n, k, prec):
     k2 = k//k1 # k2 = p^e
     v = 1 - 24*n
     pi = mpf_pi(prec)
-
```
Comment (oscarbenjamin) [inline sympy/ntheory/partitions_.py:55]: Why has this unrelated blank line been deleted?

**Comment by oscarbenjamin** [comment]
If this fixes the issue then it would be good to add tests for some problematic cases.

File: sympy/ntheory/partitions_.py [hunk 2]
```diff
@@ -172,7 +171,7 @@ def npartitions(n, verbose=False):
         math.log(10, 2)
     prec = p = int(pbits*1.1 + 100)
     s = fzero
-    M = max(6, int(0.24*n**0.5 + 4))
+    M = math.ceil(n**0.5)
```
Comment (oscarbenjamin) [inline sympy/ntheory/partitions_.py:174]: This change doesn't seem to match the description in the OP "Takes more terms in the series approximation and increases the arithmetic precision." although to be clear I haven't read the issue in detail and don't know exactly what this does

Comment (maciejskorski) [reply sympy/ntheory/partitions_.py:174]: This change takes more terms in the series approximation.

Comment (oscarbenjamin) [reply sympy/ntheory/partitions_.py:174]: I'm not going to read references etc so I'll leave this for someone else to review.

What I will say though is that I would use an integer root function rather than floating point here.

**Comment by Pontus-vB** [comment]
Maybe remove the docstring line about testing to $10^{10}$? (Apparently, not even the current version was tested properly.)

https://github.com/sympy/sympy/blob/029231be7d92bfe672dd19b1c9dd6b5a71494e2d/sympy/ntheory/partitions_.py#L145

**Comment by maciejskorski** [comment]
> Maybe remove the docstring line about testing to 1010? (Apparently, not even the current version was tested properly.)

Good point.

I am also going to use the known (tabularized) values to add a few tests.

**Comment by oscarbenjamin** [comment]
> Maybe remove the docstring line about testing to $10^{10}$? (Apparently, not even the current version was tested properly.)

It's possible that it was working and was tested extensively before but then was changed at some point and not retested. Might be worth checking back through the git history.

**Comment by jksuom** [comment]
> It's possible that it was working and was tested extensively before but then was changed at some point and not retested. 

It does not look like it was tested thoroughly. The first implementation https://github.com/sympy/sympy/commit/f008b85b74e2ff95dd956fe3817c0b93101dfb57, where testing in mentioned, says that it was tested for `10**n` up to n = 8. Then, in commit https://github.com/sympy/sympy/commit/b65d35d0206a9c654391aa5a5aa7f6961f60a71b, that was changed to n = 10, presumably meaning that `10**9` and `10**10` had been tested. Finally, in commit https://github.com/sympy/sympy/commit/101caa94c2e0f12b3dd22eab5db4a0f0d32ec902, the wording was changed to "tested through 10**10". The only bigger change in between was https://github.com/sympy/sympy/commit/b65d35d0206a9c654391aa5a5aa7f6961f60a71b where the computation of the arithmetic factor D(n) was replaced, but the new implementation looks good to me.

**Comment by oscarbenjamin** [comment]
I've made the title more descriptive.

**Comment by jksuom** [comment]
I think that the code correctly implements Rademacher's series as presented in e.g. Fredrik Johansson's article [[Joh]](https://arxiv.org/pdf/1205.5991.pdf). The series is convergent and will compute the value of `p(n)` if sufficiently many terms are taken in a partial sum.

The current number of terms `M = max(6, int(0.24*sqrt(n)+4))` was introduced in the very first commit ed221d8. There does not seem to be any explanation. My guess is that the number was determined by experimenting with some values of `n`.

The proposed value `M = math.ceil(n**0.5)` will probably suffice in a great number of cases, maybe for all values of `n` where it is feasible to perform the summation. However, a smaller number might usually suffice. For example, if `n = 10001` then the number of terms will be 100 of which the last 50 or so are rather small:
```
In [7]: npartitions(10001, verbose=True)
...
step 91 of 101 18.52438924 1.648264316e-5
step 92 of 101 0.0 1.581063201e-5
step 93 of 101 5.497489312 1.517677588e-5
step 94 of 101 12.23760173 1.457834464e-5
step 95 of 101 0.0 1.401284097e-5
step 96 of 101 2.707243334 1.347797754e-5
step 97 of 101 0.0 1.297165664e-5
step 98 of 101 0.0 1.249195209e-5
step 99 of 101 0.0 1.203709304e-5
step 100 of 101 0.0 1.160544955e-5
```
The k'th term of Rademacher's series is the product of the two numbers given on the k'th line: the arithmetic factor $A_k(n)$ that involves a Dedekind sum ("inner sum") and the steadily decreasing analytic factor $D_k(n)$.

A reliable estimate for the required number of terms can be obtained by means of Rademacher's bound M(n, N) for the remainder after a partial sum of N terms ([[Joh]](https://arxiv.org/pdf/1205.5991.pdf), (1.8)). If the remainder is less than 0.5, then the rounded partial sum should give the correct result.
```
import math

c1 = 44*math.pi**2/(225*math.sqrt(3))
c2 = math.pi*math.sqrt(2)/75
c3 = math.pi*math.sqrt(2/3)

def M(n, N):
    sqrt = math.sqrt
    return c1/sqrt(N) + c2*sqrt(N/(n - 1))*math.sinh(c3*sqrt(n)/N)
```
For n = 10001, we get
```
n [9]: M(10001, 100)
Out[9]: 0.14971976949193727

In [10]: M(10001, 50)
Out[10]: 0.5117260418430161

In [11]: M(10001, 51)
Out[11]: 0.4794655398557221
```
So 100 terms will definitely suffice, 50 terms might also work (the bound is not sharp) but, to be on the safe side, a few more terms should probably be taken.

Rademacher's bound could be used in the code to dynamically determine a sufficient number of terms.

File: sympy/ntheory/partitions_.py [hunk 3]
```diff
@@ -170,10 +170,36 @@ def npartitions(n, verbose=False):
...
+    def M(n, N):
+        sqrt = math.sqrt
+        return c1/sqrt(N) + c2*sqrt(N/(n - 1))*math.sinh(c3*sqrt(n)/N)
+    big = max(9, math.ceil(n**0.5))  # should be too large (for n > 65, ceil should work)
+    assert M(n, big) < 0.5  # else double big until too large
+    while big > 40 and M(n, big) < 0.5:
+        big //= 2
+    small = big
+    big = small*2
+    while big - small:
```
Comment (smichr) [inline sympy/ntheory/partitions_.py:189]:
```suggestion
    while big - small > 1:
```

**Comment by smichr** [comment]
`partitions` is faster; [see comment](https://github.com/sympy/sympy/issues/24267#issuecomment-1783602173)

**Comment by asmeurer** [comment]
The issue mentions testing this against the [10000 values](https://github.com/sympy/sympy/issues/24267#issuecomment-1319084927) on the OEIS. That should be done, at least manually.

**Comment by asmeurer** [comment]
The [docstring says](https://github.com/sympy/sympy/pull/24293/files#diff-44d47fd2ed268e6866e1fc65ce22dcc3740f7141f0e0d88e67ed78fe1aa9b9d4R145) the correctness has been tested up to 10**10. What was the basis for that? Is it still correct? How do you test the correctness of this function?

**Comment by smichr** [comment]
I am no longer aware of the basis for the claim but suspect that it was based on the method of summation provided that the stopping point was detected properly. I have now dynamically calculated the stopping point past which we know the error will be less than .5 so that will no longer be a source of error.

**Comment by smichr** [comment]
I ran the test comparing `npartitions` and `partition` (which uses an exact calculation) and it gave the same answer for n in `range(1000, 10_000)`. Regarding speed and memory: `partition` grows a cache of length equal to `n` while `npartitions` uses a fixed `2*10**5` elements to calculate for any `n`. The time to compute about 70 more elements for `partition` is about the same as the time needed to calculate the a value via `npartitions` for `n` approx 10000. So if we wanted to keep the cache from `partition` from growing beyond `2*10**5` we could use something like the following in `npartitions` to take advantage of the caching in `partition`:
```python
    if n <= 200_000 and n - len(_npartition) < 70 or len(_npartition) == 2 and n < 14_400:
        # There will be 2*10**5 elements created here
        # and n elements created by partition, so in case we
        # are going to be working with small n, we just
        # use partition to calculate (and cache) the values
        # since lookup is used there while summation, using
        # _factor and _totient, will be used below. But we
        # only do so if n is relatively close to the length
        # of the cache since doing 1 calculation here is about
        # the same as adding 70 elements to the cache. In addition,
        # the startup here costs about the same as calculation the first
        # 14,400 values via partition, so we delay startup here unless n
        # is smaller than that.
        return partition(n)
```
Adding this allows the first 10k partitions to be computed about 20X faster.

File: sympy/ntheory/partitions_.py [hunk 4]
```diff
@@ -157,11 +157,27 @@ def npartitions(n, verbose=False):
...
+        # There will be 2*10**5 elements created here
+        # and n elements created by partition, so in case we
+        # are going to be working with small n, we just
+        # use partition to calculate (and cache) the values
+        # since lookup is used there while summation, using
+        # _factor and _totient, will be used below. But we
+        # only do so if n is relatively close to the length
+        # of the cache since doing 1 calculation here is about
+        # the same as adding 70 elements to the cache. In addition,
+        # the startup here costs about the same as calculation the first
```
Comment (smichr) [inline sympy/ntheory/partitions_.py:177]:
```suggestion
        # the startup here costs about the same as calculating the first
```

Comment (smichr) [reply sympy/ntheory/partitions_.py:177]: I'll wait to commit this change until other review is done.

File: sympy/ntheory/partitions_.py [hunk 5]
```diff
@@ -170,10 +186,36 @@ def npartitions(n, verbose=False):
...
+        N = (big + small)//2
+        if (er := M(n, N)) < 0.5:
+            big = N
+        elif er >= 0.5:
+            small = N
+    M = big  # done with function M; now have value
+
+    # sanity check for expected size of answer
     if M > 10**5:
         raise ValueError("Input too big") # Corresponds to n > 1.7e11
```
Comment (smichr) [inline sympy/ntheory/partitions_.py:215]:
```suggestion
    if M > 10**5:  # i.e. M > maxn
        raise ValueError("Input too big")  # i.e. n > 149832547102
```

**Comment by smichr** [comment]
The largest n that can be computed for a given M is shown below (as computed from the Rademacher's bound
```
M             n
100000	149832547102
10000	1129559468
1000	7978280
100	48856
10	81
```
